### PR TITLE
PriceAuthority for multipoolAutoswap

### LINKED ISSUE #1947: Add a PriceAuthority to Autoswap and Multipool Autoswap. Retain old methods for backwards compatiblity
**What is the Problem Being Solved?**

We will need to use Autoswap or Multipool Autoswap in the loan contract for liquidation. It would be nice if they also included the priceAuthority API such that getting quotes was standardized. 

Additionally, Autoswap and Multipool Autoswap may give slightly different quotes based on whether you ask `getInputPrice` or `getOutputPrice`. From the perspective of the `priceAuthority`, if the quotes are indistinguishable, the quoted amountOut should be also the same. We should fix this in Autoswap and Multipool Autoswap and write tests to ensure the quoted amount is the same if the quotes are otherwise indistinguishable. 

**Description of the Design**

We should add a priceAuthority but still retain the methods `getInputPrice` and `getOutputPrice` for backwards compatibility.

**Security Considerations**

**Test Plan**

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
closes #1947 

PriceAuthority for multipoolAutoswap built on top of PriceAuthority.

Includes adding a notifier that gets updated whenever the pool balances change. The priceAuthority uses that to trigger checking the price levels.

### CODE REVIEW

File: packages/zoe/src/contracts/multipoolAutoswap/pool.js [hunk 1]
```diff
@@ -170,6 +190,7 @@ export const makeAddPool = (zcf, isSecondary, initPool, centralBrand) => {
           'insufficient Secondary deposited',
         );
 
+        updateState(pool);
```
Comment (katelynsills) [inline packages/zoe/src/contracts/multipoolAutoswap/pool.js]: I think this is being updated twice for the same action.

File: packages/zoe/src/contracts/multipoolAutoswap/types.js [hunk 2]
```diff
@@ -80,4 +84,9 @@
  * of amountOut that can be gained for that amountIn.
  * @property {(brand: Brand) => Record<string, Amount>} getPoolAllocation get an
  * AmountKeywordRecord showing the current balances in the pool for brand.
+ * getQuoteIssuer {() => Issuer} getQuoteIssuer - get the Issuer that attests to
+ * the prices in the priceQuotes issued by the PriceAuthorities
+ * @property {(brand: Brand) => Record<string, PriceAuthority>} getPriceAuthorities
```
Comment (katelynsills) [inline packages/zoe/src/contracts/multipoolAutoswap/types.js]:
```suggestion
 * @property {(brand: Brand) => {toCentral: PriceAuthority, fromCentral: PriceAuthority}} getPriceAuthorities
```

**Review by katelynsills — APPROVED** [review]
LGTM!

File: packages/zoe/src/contracts/multipoolAutoswap/pool.js [hunk 3]
```diff
@@ -215,9 +235,38 @@ export const makeAddPool = (zcf, isSecondary, initPool, centralBrand) => {
...
+      pool.getPriceGivenAvailableInput,
+      secondaryBrand,
+      centralBrand,
+      timer,
+      zcf,
+      notifier,
+      quoteIssuerKit,
+    );
+    const fromCentralPriceAuthority = makePriceAuthority(
+      pool.getPriceGivenAvailableInput,
```
Comment (michaelfig) [inline packages/zoe/src/contracts/multipoolAutoswap/pool.js:273]: Out of ignorance, should both of these `makePriceAuthority` calls' first arguments (`pool.getPriceGivenAvailableInput`) be the same?

File: packages/zoe/src/contracts/multipoolAutoswap/priceAuthority.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    const quoteAmount = amountMath.make(quote, brand);
+    const quotePayment = quoteIssuerKit.mint.mintPayment(quoteAmount);
+    return harden({ quoteAmount, quotePayment });
+  };
+
+  // called by createQuote for each trigger to compare price to trigger level
+  const calcAmountOut = amountIn =>
+    getOutputForGivenInput(amountIn, actualBrandOut).amountOut;
+  const calcAmountIn = amountOut =>
+    getOutputForGivenInput(amountOut, actualBrandIn).amountIn;
```
Comment (michaelfig) [inline packages/zoe/src/contracts/multipoolAutoswap/priceAuthority.js:33]: Again, should both of these `getOutputForGivenInput` calls use the same function?

File: packages/zoe/src/contracts/multipoolAutoswap/types.js [hunk 5]
```diff
@@ -43,6 +43,10 @@
  * @property {() => ZCFSeat} getPoolSeat
  * @property {() => Amount} getSecondaryAmount
  * @property {() => Amount} getCentralAmount
+ * @property {() => Notifier} getNotifier
+ * @property {() => Void} updateState
```
Comment (michaelfig) [inline packages/zoe/src/contracts/multipoolAutoswap/types.js]:
```suggestion
 * @property {() => void} updateState
```

File: packages/zoe/src/contracts/multipoolAutoswap/types.js [hunk 6]
```diff
@@ -80,4 +84,9 @@
  * of amountOut that can be gained for that amountIn.
  * @property {(brand: Brand) => Record<string, Amount>} getPoolAllocation get an
  * AmountKeywordRecord showing the current balances in the pool for brand.
+ * getQuoteIssuer {() => Issuer} getQuoteIssuer - get the Issuer that attests to
```
Comment (michaelfig) [inline packages/zoe/src/contracts/multipoolAutoswap/types.js]:
```suggestion
 * @property {() => Issuer} getQuoteIssuer - get the Issuer that attests to
```

**Review by michaelfig — APPROVED** [review]
Looking good, just want extra attention where I called on you.

File: packages/zoe/src/contracts/multipoolAutoswap/priceAuthority.js (hunk 4 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/multipoolAutoswap/priceAuthority.js:33]: I think so, but the fact that calcAmountIn isn't actually used (`priceAuthority.js` lines 106 and 169) may be leaving me confused about the intended semantics.

File: packages/zoe/src/contracts/multipoolAutoswap/types.js (hunk 5 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/multipoolAutoswap/types.js]: done.  Why does typescript let me get away with this?

File: packages/zoe/src/contracts/multipoolAutoswap/types.js (hunk 6 above)
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/multipoolAutoswap/types.js]: done.

File: packages/zoe/src/contracts/multipoolAutoswap/priceAuthority.js (hunk 4 above)
Comment (michaelfig) [reply packages/zoe/src/contracts/multipoolAutoswap/priceAuthority.js:33]: `calcAmountIn` is used: I think you missed `priceAuthority.js` line 257.

Comment (michaelfig) [reply packages/zoe/src/contracts/multipoolAutoswap/priceAuthority.js:33]: I think the `calcAmountIn` is incorrect.  It should not calculate output, it should calculate input.  I'm fixing it the next PR.

**Pull request merged**